### PR TITLE
fetch_key_from_memc is nil out of do block

### PR DESCRIPTION
**Pull request by xiewenhui**
fetch_key_from_memc is nil out of do block

### CODE REVIEW

**Comment by xiewenhui** [comment]
fix init bug: mv do block to init()

File: lualib/ngx/ssl/session/ticket/key_rotation.lua [hunk 1]
```diff
@@ -302,6 +261,51 @@ function _M.init(opts)
...
+            memc_fetch_retry_delay = memc_fetch_retry_delay,
+
+            memc_conn_max_idle_time = memc_conn_max_idle_time,
+
+            memc_store_retries = 0,
+            memc_store_retry_delay = 0,
+        }
+    end
+
+
```
Comment (agentzh) [inline lualib/ngx/ssl/session/ticket/key_rotation.lua]: Extra blank line. We only need a single blank line here.

**Review by agentzh — CHANGES_REQUESTED** [review]

**Comment by agentzh** [comment]
Despite the minor coding style thing mentioned above, this looks good to me.

@doujiang24 Will you have a quick look at this?

File: lualib/ngx/ssl/session/ticket/key_rotation.lua [hunk 2]
```diff
@@ -57,52 +57,15 @@ end
 
 local shdict_name, shm_cache_pos_ttl, shm_cache_neg_ttl, disable_shm_cache
```
Comment (doujiang24) [inline lualib/ngx/ssl/session/ticket/key_rotation.lua]: These variables seems no need to stay in the top level, maybe we can just move them to `init` function?

Comment (doujiang24) [reply lualib/ngx/ssl/session/ticket/key_rotation.lua]: And we should add `local disable_shm_cache = opts.disable_shm_cache` in `init`.

Comment (xiewenhui) [reply lualib/ngx/ssl/session/ticket/key_rotation.lua]: umm，some top level variables can move  to init() func,  @agentzh @doujiang24

Comment (xiewenhui) [reply lualib/ngx/ssl/session/ticket/key_rotation.lua]: @agentzh @doujiang24 i made some changes based on your comments.

File: lualib/ngx/ssl/session/ticket/key_rotation.lua [hunk 3]
```diff
@@ -283,25 +238,68 @@ function _M.init(opts)
...
 
-    memc_conn_max_idle_time = opts.memc_conn_max_idle_time
+    local memc_conn_max_idle_time = opts.memc_conn_max_idle_time
 
     local frandom = assert(io.open("/dev/urandom", "rb"))
     fallback_random_key = frandom:read(48)
     frandom:close()
 
+    nkeys = floor(ticket_ttl / time_slot) + 2
+    do
```
Comment (doujiang24) [inline lualib/ngx/ssl/session/ticket/key_rotation.lua:261]: Seems the `do end` is not so necessary here? The local variables can be auto GCed after the function finished.

File: lualib/ngx/ssl/session/ticket/key_rotation.lua [hunk 4]
```diff
@@ -283,25 +238,68 @@ function _M.init(opts)
...
+            memc_fetch_retry_delay = memc_fetch_retry_delay,
+
+            memc_conn_max_idle_time = memc_conn_max_idle_time,
+
+            memc_store_retries = 0,
+            memc_store_retry_delay = 0,
+        }
+    end
+
     local ctx = {}
```
Comment (doujiang24) [inline lualib/ngx/ssl/session/ticket/key_rotation.lua:303]: Seems the `ctx` in the whole file is useless? remove it also?

Comment (xiewenhui) [inline lualib/ngx/ssl/session/ticket/key_rotation.lua]: should not,  ctx is used in fetch_key_from_memc func implementation for store memcached obj created

File: lualib/ngx/ssl/session/ticket/key_rotation.lua (hunk 3 above)
Comment (xiewenhui) [reply lualib/ngx/ssl/session/ticket/key_rotation.lua:261]: actually not so necessary,  but it makes the  initialization more structured and clear.  i prefer to keep 'do ... end' here

File: lualib/ngx/ssl/session/ticket/key_rotation.lua (hunk 4 above)
Comment (xiewenhui) [reply lualib/ngx/ssl/session/ticket/key_rotation.lua:303]: should not, ctx is used in fetch_key_from_memc func implementation for store memcached obj created (for efficiency, reduce temp obj create)

Comment (doujiang24) [reply lualib/ngx/ssl/session/ticket/key_rotation.lua:303]: @xiewenhui Yeah, you are right:) Seems useless only in a few functions like `dlog`, but it seems doesn't matter.

File: lualib/ngx/ssl/session/ticket/key_rotation.lua (hunk 3 above)
Comment (doujiang24) [reply lualib/ngx/ssl/session/ticket/key_rotation.lua:261]: Yeah, I'm fine with keep it :)

**Comment by agentzh** [comment]
Merged. Thanks!

**Pull request closed without merging**